src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc

```diff
@@ -23,153 +23,190 @@ namespace ray {
+bool HybridSchedulingPolicy::IsNodeFeasible(
+    const scheduling::NodeID &node_id,
+    const NodeFilter &node_filter,
+    const NodeResources &node_resources,
+    const ResourceRequest &resource_request) const {
+  if (!is_node_alive_(node_id)) {
+    return false;
   }
 
-  const auto start_index = round.size();
-  for (const auto &pair : nodes_) {
-    if (pair.first != preferred_node_id &&
-        predicate(pair.first, pair.second.GetLocalView())) {
-      round.push_back(pair.first);
+  if (node_filter != NodeFilter::kAny) {
+    const bool has_gpu = node_resources.total.Has(ResourceID::GPU());
+    if (node_filter == NodeFilter::kGPU && !has_gpu) {
+      return false;
+    } else if (node_filter == NodeFilter::kNonGpu && has_gpu) {
+      return false;
     }
   }
-  // Sort all the nodes, making sure that if we added the local node in front, it stays in
-  // place.
-  std::sort(round.begin() + start_index, round.end());
-
-  scheduling::NodeID best_node_id = scheduling::NodeID::Nil();
-  float best_utilization_score = INFINITY;
-  bool best_is_available = false;
-
-  // Step 2: Perform the round robin.
-  auto round_it = round.begin();
-  for (; round_it != round.end(); round_it++) {
-    const auto &node_id = *round_it;
-    const auto &it = nodes_.find(node_id);
-    RAY_CHECK(it != nodes_.end());
-    const auto &node = it->second;
-    if (!node.GetLocalView().IsFeasible(resource_request)) {
-      continue;
-    }
 
-    bool ignore_pull_manager_at_capacity = false;
-    if (node_id == local_node_id_) {
-      // It's okay if the local node's pull manager is at
-      // capacity because we will eventually spill the task
-      // back from the waiting queue if its args cannot be
-      // pulled.
-      ignore_pull_manager_at_capacity = true;
-    }
-    bool is_available = node.GetLocalView().IsAvailable(resource_request,
-                                                        ignore_pull_manager_at_capacity);
-    float critical_resource_utilization =
-        node.GetLocalView().CalculateCriticalResourceUtilization();
-    RAY_LOG(DEBUG) << "Node " << node_id.ToInt() << " is "
-                   << (is_available ? "available" : "not available") << " for request "
-                   << resource_request.DebugString()
-                   << " with critical resource utilization "
-                   << critical_resource_utilization << " based on local view "
-                   << node.GetLocalView().DebugString();
-    if (critical_resource_utilization < spread_threshold) {
-      critical_resource_utilization = 0;
-    }
+  return node_resources.IsFeasible(resource_request);
+}
 
-    bool update_best_node = false;
+namespace {
+float ComputeNodeScore(const NodeResources &node_resources, float spread_threshold) {
+  float critical_resource_utilization =
+      node_resources.CalculateCriticalResourceUtilization();
+  if (critical_resource_utilization < spread_threshold) {
+    critical_resource_utilization = 0;
```

Comment:
yeah I agree this is something we can potentially improve; but we need a bit more benchmark to see the side effect of changing this behavior. in theory removing this could lead to node more overloaded